Count store hanging and inconsistent state after rotation

MishaDemianenko opened the pull request:
Revert changes introduced to ConcurrentMapState that cause exception
on updates with old versions.
Prevent cases when count store rotation causing TransactionRepresentationCommitProcess to hang.
Hang scenario:

```
+------------------------+---------------------------------+
| T1:Checkpointer        |       T2:CommitProcess          |
+----------------------------------------------------------+
|                        |publish transaction as commited  |
|                        |                                 |
|   rotation             |                                 |
|                        |                                 |
|                        |apply transaction to the store   |
|                        |                                 |
+------------------------+---------------------------------+
```

On the last step of T2 exception will be raised. And in case if transaction contains legacy indexes
that will prevent closing LegacyIndexApplier and removing current transactionId from
transactionOrdering queue. As result all new transaction will not be able to appendToLog
since their will wait forever on orderLegacyIndexChanges.

MishaDemianenko committed:
ConcurrentMapState should allow and ignore updates for versions less or equal to current

Revert changes introduced to ConcurrentMapState that cause exception
on updates with old versions.
Prevent cases when count store rotation causing TransactionRepresentationCommitProcess to hang.
Hang scenario:
+------------------------+---------------------------------+
| T1:Checkpointer        |       T2:CommitProcess          |
+----------------------------------------------------------+
|                        |publish transaction as commited  |
|                        |                                 |
|   rotation             |                                 |
|                        |                                 |
|                        |apply transaction to the store   |
|                        |                                 |
+------------------------+---------------------------------+
On the last step of T2 exception will be raised. And in case if transaction contains legacy indexes
that will prevent closing LegacyIndexApplier and removing current transactionId from
transactionOrdering queue. As result all new transaction will not be able to appendToLog
since their will wait forever on orderLegacyIndexChanges.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/ConcurrentMapState.java:
```diff
@@ -73,12 +66,10 @@ public String toString()
         {
-            throw new IllegalStateException( "Cannot apply update with given version " + version +
-                                             " when base version is " + previousVersion );
+            return EntryUpdater.noUpdates();
```
Not convinced by this change. This should never happen and in this way we can hide bugs in other parts of the code.

MishaDemianenko replied:
Our count store rotation based on last committed transaction and since we publish transaction as committed before apply transaction to count store that _can_ be the case

davidegrohmann replied:
That should never happen and this change will hide that problem.

MishaDemianenko replied:
How that can never happen if there is no sync between those 2?

davidegrohmann replied:
There is, this code should do it https://github.com/neo4j/neo4j/blob/2.3/community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/RotationState.java#L69 If it does not, then we have a bug there.

MishaDemianenko replied:
Those are based on committed transactions right? and those published as committed before applying changes.

MishaDemianenko committed:
Counts store rotation post state changes and applied transactions tracking

Rotation post state should apply all transactions that updating count store
but only transaction with ids that greater than rotation version should be treated
as applied in post state.
Otherwise during next count store rotation those transactions will be incorectly interpreted
as already applied during current count store rotation cycle.

MishaDemianenko commented:
@davidegrohmann , @tinwelint update PR to include fix for post state applied transaction counting

davidegrohmann commented:
@MishaDemianenko btw this PR should be against 2.2

MishaDemianenko commented:
@davidegrohmann since we do not have check pointer thread in 2.2 and should not have those problems there should be fine to have it only in 2.3

MishaDemianenko commented:
@thobe can you please also take a look

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/ConcurrentMapState.java:
```diff
@@ -73,19 +70,19 @@ public String toString()
         }
-        versionSequence.offer( version, META );
         update( highestAppliedVersion, version );
+        hasTrackedChanges.set( true );
```
Just to check if my understanding is correct, is this necessary to make sure we return true from `hasChanges()` even before updating the `changes` map?

MishaDemianenko replied:
we need both like `hasTrackedChanges.get() && !changes.isEmpty()` - so changes map should be not empty and it should contain at least one tracked change

davidegrohmann replied:
ok, so I am not sure I understand this, why do you need `hasTrackedChanges`?

MishaDemianenko replied:
`hasTrackedChanges` is used to determine  if we have any changes that satisfy `version > threshold` - meaning changes that came in a post state during rotation. And as soon as we have at least one - our state hasChanges. If we do not have any - all changes where from previous state

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/TransactionRepresentationCommitProcessIT.java:
```diff
@@ -0,0 +1,427 @@
+    @After
+    public void tearDown()
+    {
+        lifeRule.shutdown();
```
I don't think this is necessary

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/TransactionRepresentationCommitProcessIT.java:
```diff
@@ -0,0 +1,427 @@
+
+        private void randomSleep() throws InterruptedException
+        {
+            Thread.sleep( new Random( System.currentTimeMillis() ).nextInt( 50 ) );
```
`ThreadLocalRandom.current()` instead?

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/kvstore/AbstractKeyValueStoreTest.java:
```diff
@@ -275,54 +217,15 @@ public void write( WritableBuffer key, WritableBuffer value )
-            }
-        };
+        final Store store = resourceManager.managed( createTestStore() );
+        updateStore( store, 1 );
```
:+1:

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/kvstore/ConcurrentMapStateTest.java:
```diff
@@ -133,52 +132,18 @@ public void shouldCreateAnUpdaterForMultipleVersionUpdatesNotInOrder() throws Ex
 
     @Test
-    public void shouldFailIfApplyingAVersionUpdateWithAVersionLowerOrEqualToTheInitialVersion() throws Exception
+    public void shouldUseEmptyUpdaterOnVersionLowerOrEqualToTheInitialVersion() throws Exception
```
Not sure about this one, see other comment.

MishaDemianenko replied:
yep that we have talk on top of PR, but for me it still sound like creating empty updater is still better in that case comparing to throwing exception

MishaDemianenko commented:
@davidegrohmann fix for not commented comments :)

thobe commented:
looks mostly ok.

davidegrohmann commented:
@MishaDemianenko there a couple of required small fixes pending here

tinwelint commented:
@MishaDemianenko have all issues been resolved here? I don't understand all implications of these fixes, but I can't spot any obvious problem either.

MishaDemianenko commented:
@tinwelint yes all issues resolved and all questions answered
